### PR TITLE
Design for New Saved Object Service Interface for Custom Repository

### PR DESCRIPTION
**Pull request by bandinib-amzn**
### Description

RFC for new saved object service interface for custom repository
[Rendered RFC](https://github.com/opensearch-project/OpenSearch-Dashboards/blob/daf7b91302c3814f703f98b986ad39a2147d0265/docs/saved_objects/saved_object_repository_factory_design.md)

### Check List

- [ ] All tests pass
  - [ ] `yarn test:jest`
  - [ ] `yarn test:jest_integration`
  - [ ] `yarn test:ftr`
- [ ] New functionality includes testing.
- [ ] New functionality has been documented.
- [ ] Update [CHANGELOG.md](./../CHANGELOG.md)
- [ ] Commits are signed per the DCO using --signoff

### CODE REVIEW

File: docs/saved_objects/saved_object_repository_factory_design.md [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+        * Some components of Saved object service such as Serializer, Type registry, interface to create internal and scoped repository are only available during Saved Object Service Start. As per the current architecture, first Saved Object Service Setup → Plugin Setup → Saved Object Service Start → Plugin Start. Some core plugin (e.g. opensearch_dashboards_usage_collection) calls find operation before plugin start and it fails because some components are still not available before plugin start.  
+
+- **Approach 3 (Preferred)**: The proposed architecture will add one more layer of abstraction in Saved Object Service. `The Repository Factory Provider` in Dashboards will be responsible for creating and managing instances of the Repository (e.g. SavedObjectRepository, PostgresRepository, DynamoDBRepository etc.), which is used to interact with the metadata storage that stores the saved objects. Currently we have an repository interface named `ISavedObjectRepository`, and the `SavedObjectRepository` is the implementation, which use an OpenSearch index as its data store. This approach would make the implementation of `ISavedObjectRepository` replaceable by plugin.
```
Comment (zhongnansu) [inline docs/saved_objects/saved_object_repository_factory_design.md:47]: Better add code link to the existing component, using markdown

File: docs/saved_objects/saved_object_repository_factory_design.md [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+        typeRegistry,
+        serializer,
+        migrator,
+        allowedTypes,
+    });
+    }
+    ```
+2. Register the provider: Register the repository factory provider with right dependencies. 
+    ```
+    core.savedObjects.registerRepositoryFactoryProvider(repositoryFactoryProvider);
```
Comment (zhongnansu) [inline docs/saved_objects/saved_object_repository_factory_design.md:88]: use `portgresRepositoryFactory` for naming to better callout the usage?

**Comment by ashwin-pc** [comment]
I like the approach but if approach 3 is preferred, can you move it higher up in the lists and move the other two to alternatives considered? Since you aren't providing implementation details for the other two.

Also I didn't understand the cons for the preferred approach. What do you mean by code reusability is low?

And I know this is and rfc and not a design, but is there plan for a contract that new saved object repository needs to adhere to? Like features it needs to meet. And have you looked to see if there is any feature that we will not be able to support with some of the data stores that we are looking to support like postgress.

Nit: would Adapter be a better term here than Repository?  Since each of them act as an adapter to a new data store.

File: docs/saved_objects/saved_object_repository_factory_design.md [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+
+Currently, Dashboards stores its metadata configuration inside OpenSearch index (called .kibana). This approach is by design of Dashboards and biased towards product decision by upstream which works seamlessly and out of the box for customers but it introduces challenges while operating at scale and providing high availability for Dashboards. While choosing OpenSearch as a storage for Dashboards metadata, availability of Dashboards depends on OpenSearch cluster’s availability and other cluster parameters such as cluster health, state, versions which could make Dashboards unavailable.
+
+To mitigate above problem and unblock future extensibility of Dashboards, we are building Dashboards Meta storage adaptor to decouple Dashboards metadata storage from OpenSearch. This project will focus on introducing new interface in Saved Object Service using which developer can build their custom repository and save Dashboards metadata in storage of their choice.
+
+The stakeholders of this new interface include the developers of the Dashboards, developers of the Managed Service, and community contributors who wants to use other metadata store.
```
[the remark this answers is not in the document]
Comment (bandinib-amzn) [reply docs/saved_objects/saved_object_repository_factory_design.md]: Removed the managed service from the text.

File: docs/saved_objects/saved_object_repository_factory_design.md [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+
+## Architecture Overview
+
+The Saved Object Service is a critical component of Dashboards that provides a way to store and manage application data. It is built using a modular architecture that provides a high degree of flexibility and extensibility. The new interface will be designed to replace [ISavedObjectRepository](https://github.com/opensearch-project/OpenSearch-Dashboards/blob/main/src/core/server/saved_objects/service/lib/repository.ts#L134) implementation so that developers can build plugins that leverage the power of existing saved object service and use their own database to store and retrieve metadata of Dashboards.
```
[the remark this answers is not in the document]
Comment (bandinib-amzn) [reply docs/saved_objects/saved_object_repository_factory_design.md]: Updated.

File: docs/saved_objects/saved_object_repository_factory_design.md (hunk 1 above)
Comment (bandinib-amzn) [reply docs/saved_objects/saved_object_repository_factory_design.md]: Good idea! I have added code links to existing components.

File: docs/saved_objects/saved_object_repository_factory_design.md (hunk 2 above)
Comment (bandinib-amzn) [reply docs/saved_objects/saved_object_repository_factory_design.md:88]: `Repository factory provider` is not limited to Postgres. PostgresRepository is just one use case.

**Comment by bandinib-amzn** [comment]
> I like the approach but if approach 3 is preferred, can you move it higher up in the lists and move the other two to alternatives considered? Since you aren't providing implementation details for the other two.

Thanks for the suggestion. I moved Approach 3 higher up in the list. I haven't provided implementation details for other 2 because they are no preferred. But I have POC for each approach and evaluated based on that. If it helps, I can attach POC link under each approach.

> Also I didn't understand the cons for the preferred approach. What do you mean by code reusability is low?

If you check current [SavedObjectRepository](https://github.com/opensearch-project/OpenSearch-Dashboards/blob/main/src/core/server/saved_objects/service/lib/repository.ts#L249-L263) implementation, each saved object operation contains validation logic before calling OpenSeach API. So the validation logic is inside the  `SavedObjectRepository`. We will need refactor saved object service in order to make those validation code accessible to plugin to reuse. One way to do that is we can move validation logic to [SavedObjectClient](https://github.com/opensearch-project/OpenSearch-Dashboards/blob/main/src/core/server/saved_objects/service/saved_objects_client.ts#L302-L441), so that any repository will go through that validation. For initial release, I am planning to focus on just adding `Repository Factory Provider` to the Saved Object Service and build external plugin (with Postgres database) as use case for that interface. 

> And I know this is and rfc and not a design, but is there plan for a contract that new saved object repository needs to adhere to? Like features it needs to meet. And have you looked to see if there is any feature that we will not be able to support with some of the data stores that we are looking to support like postgress.

Repository for saved object needs adhere to contracts defined in [ISavedObjectRepository](https://github.com/opensearch-project/OpenSearch-Dashboards/blob/main/src/core/server/saved_objects/service/lib/repository.ts#L134). See the approach 3. Please check the poc links listed below for external PostgresRepository plugin.
- [Core Dashboard changes](https://github.com/bandinib-amzn/OpenSearch-Dashboards/commit/b9cfc14a294854340d7db9de6632fe1b1cc2f6bd)
- [Postgres Repository Plugin](https://github.com/bandinib-amzn/metadata_plugin/commit/dac35f0198a54ce1e76ca8dfcba87e949d4e6dfd#diff-a3cb4bf8c509eb6160c39130fbb6271089f67dad50d28d03c094db4fec6683f1) 

We haven't evaluated all the data storage options. It is currently in progress. But looking at saved object operation (create, get, update, delete, find etc.), each operation would be possible for any type of data store. While evaluating the options, if I found any limitation, I'll definitely call out.

> Nit: would Adapter be a better term here than Repository? Since each of them act as an adapter to a new data store.

Are you talking about the name of the plugin or the repository factory provider?

**Comment by ashwin-pc** [comment]
@bandinib-amzn Thanks, FYI, Great job, I really like this simple and clean abstraction. Most of my comments are just trying to poke holes here, but i dont think there is anything wrong with the suggested approach.

> If you check current [SavedObjectRepository](https://github.com/opensearch-project/OpenSearch-Dashboards/blob/main/src/core/server/saved_objects/service/lib/repository.ts#L249-L263) implementation, each saved object operation contains validation logic before calling OpenSeach API. So the validation logic is inside the SavedObjectRepository. We will need refactor saved object service in order to make those validation code accessible to plugin to reuse. One way to do that is we can move validation logic to [SavedObjectClient](https://github.com/opensearch-project/OpenSearch-Dashboards/blob/main/src/core/server/saved_objects/service/saved_objects_client.ts#L302-L441), so that any repository will go through that validation. For initial release, I am planning to focus on just adding Repository Factory Provider to the Saved Object Service and build external plugin (with Postgres database) as use case for that interface.

I see. I think the way i see it, validation logic specific to a datasource connector should reside in the appropriate "Repository" and validation logic specific to `Saved Objects` should be pulled out and implemented at a higher abstraction layer like you mentioned the Saved Object client (although based on your comment i take it its not something you are doing right now, which is fine as long as we track it somewhere). So this isnt really a con of the approach, but a con of how you plan on implementing it. So is there really a con with this approach itself?

> I haven't provided implementation details for other 2 because they are no preferred. But I have POC for each approach and evaluated based on that. If it helps, I can attach POC link under each approach.

Yeah that would be helpful. I only asked because the preferred approach was lower down the list but the only one with implementation details. 

> Repository for saved object needs adhere to contracts defined in [ISavedObjectRepository](https://github.com/opensearch-project/OpenSearch-Dashboards/blob/main/src/core/server/saved_objects/service/lib/repository.ts#L134). See the approach 3. Please check the poc links listed below for external PostgresRepository plugin.

Meta question, prior to this change we could already create multiple repositories to interact with saved objects. Do you know if any plugin used this to create their own repository? And if so how does this change impact that? Do we expect a single instance of OSD to connect to multiple datasource types.

> Are you talking about the name of the plugin or the repository factory provider?

ignore my comment here. This was before i learnt about the repository pattern :)

**Review by zhongnansu — APPROVED** [review]

**Comment by bandinib-amzn** [comment]
> I see. I think the way i see it, validation logic specific to a datasource connector should reside in the appropriate "Repository" and validation logic specific to Saved Objects should be pulled out and implemented at a higher abstraction layer like you mentioned the Saved Object client (although based on your comment i take it its not something you are doing right now, which is fine as long as we track it somewhere). So this isnt really a con of the approach, but a con of how you plan on implementing it. So is there really a con with this approach itself?

I couldn't think anything else other than that.

> Yeah that would be helpful. I only asked because the preferred approach was lower down the list but the only one with implementation details.

I have added POC links for each approach. 

> Meta question, prior to this change we could already create multiple repositories to interact with saved objects. Do you know if any plugin used this to create their own repository? And if so how does this change impact that? 

I don't know any plugin which have created their own repository. But I know that [security-dashboards-plugin](https://github.com/opensearch-project/security-dashboards-plugin) uses [Saved Object Wrapper](https://github.com/opensearch-project/security-dashboards-plugin/blob/main/server/saved_objects/saved_objects_wrapper.ts) which is another abstraction layer in Saved Object Service. We explored `Saved Object Wrapper` to use for our case, but due to some limitations (listed under cons of Approach 2), we decided to introduce another abstraction layer. 

> Do we expect a single instance of OSD to connect to multiple datasource types.

No, we don't expect a single instance of OSD to connect to multiple datasource types. Do you have any use case where we want single instance of OSD to connect to multiple datasource.

**Review by zhongnansu — APPROVED** [review]

**Review by ashwin-pc — CHANGES_REQUESTED** [review]
@bandinib-amzn Thanks for addressing the comments. I think the proposal is great! but dont think that this PR should be merged though. Blocking it so that we dont accidentally merge it in until we discuss that among the maintainers and come to a consensus.

**Comment by zhongnansu** [comment]
> @bandinib-amzn Thanks for addressing the comments. I think the proposal is great! but dont think that this PR should be merged though. Blocking it so that we dont accidentally merge it in until we discuss that among the maintainers and come to a consensus.

@ashwin-pc Hi Ashwin, do you want to list all your concerns here for maintainers to discuss? From my side, the PR is good to go. Consider it as a high-level-design, that we want to keep in our repo for version control. Same as we did for other projects, e.g. https://github.com/opensearch-project/OpenSearch-Dashboards/blob/main/docs/multi-datasource/high_level_design.md

**Comment by bandinib-amzn** [comment]
> @ashwin-pc Hi Ashwin, do you want to list all your concerns here for maintainers to discuss? From my side, the PR is good to go. Consider it as a high-level-design, that we want to keep in our repo for version control. Same as we did for other projects, e.g. https://github.com/opensearch-project/OpenSearch-Dashboards/blob/main/docs/multi-datasource/high_level_design.md

+1

**Comment by ashwin-pc** [comment]
@zhongnansu Sure :) didnt want to pollute the comment thread to discuss that since we are also discussing the proposal in general here, but here are my thoughts and concerns. 

1. Merging a pull request before implementing the actual feature can cause confusion about whats a feature and whats a proposal. 
2. Since the final implementation can differ from a proposal, this can unnecessarily pollute the version control history, be out of sync with the final implementation or both. 
3. Changes that go into the repo are either functional code or documentation about that functional code. This PR is neither. It is desirable to keep the codebase clean and by adding things that dont relate to either one is IMO unnecessary. 
4. Yes, Git does offer version control, but we already have tools like GitHub that offer the same historical information, with additional context about the discussions that followed. The discussions are lost in code. 

> Consider it as a high-level-design, that we want to keep in our repo for version control. Same as we did for other projects, e.g. https://github.com/opensearch-project/OpenSearch-Dashboards/blob/main/docs/multi-datasource/high_level_design.md

This isnt the same as the one you mentioned. That was the design for an implemented feature. I'm totally okay merging this once the final feature is implemented and the PR still accurately reflects the implementation, my concern is merging it before the implementation is complete. 

However if there are any other reasons other than version control for merging this in before the implementation is complete, id love to hear that.

**Comment by seraphjiang** [comment]
@ashwin-pc 
Design, Testing, Doc are all part of code. There should not be any different.

I agree, there is no one time design, one time code. The implementation could change moving forward, even design could has bug How we handle design bug/enhancement, it should be same as how we treat other code defect. File an issue and address it with another PR. 

MDS design is not a good example reference, we add design after implementation. Design should always ahead of implementation. Design also need iteration. 

Design/Architecture could be out-of-date, we should also allow to remove deprecate design just like deprecate code.


For this interface design, it is important change on lower level, the implementation except interface/abstract won't necessary be part of OSD core code base. We should keep this transparent to all OSD dev/Plugin dev , make it easy to understand the change and easy to access the information even without search github issue.

**Comment by ashwin-pc** [comment]
>  Design, Testing, Doc are all part of code. There should not be any different.

Yep, but this is none of the above, this is an RFC/proposal. A proposal is the "intended" design. all the others are added either during or after implementation.

> I agree, there is no one time design, one time code.

@seraphjiang I'm confused. I never said there is no one time design. I only said that a proposal is different from the final implemented design and i'm all for adding the final implemented design in the code, just not the proposal.

> Design/Architecture could be out-of-date, we should also allow to remove deprecate design just like deprecate code.

Exactly! And they are usually removed alongside the change that makes them out of date. Yes we do remove them out of turn sometimes, but thats not a practice we should emulate. That's why in the PR template we have a checkbox asking the contributor if the docs for the corresponding change have been updated. Docs and change should go in together. But if they cant, the change should go in first followed by the docs. Think of it like the changelog. We dont add the changelog entry until the change itself is committed.

> Design should always ahead of implementation. Design also need iteration.

I agree, but that design needs discussion and iteration, none of which code is the best place to do. Which is why we follow the current structure:

1. RFC/Proposal: submit a proposal as a Github issue
2. Discussion: Everyone can discuss using Github comments on a public forum and agree on the solution
4. Implement: Once agreed, the feature and its corresponding documentation can be raised in another PR. 
5. Review and merge: The new PR and its docs are reviewed against the agreed upon design and merged and the original RFC/Proposal issue can be closed

And again, to highlight, The title of the PR states that this is an RFC (Request for Comment). If that's the case why are we calling it a design, and now debating that we include it in code? And if that's a typo and this is indeed the HLD, then whats the harm in waiting until the implementation is complete? Who will benefit from having this in code base before the implementation is ready?

**Comment by seraphjiang** [comment]
I don't think this is proposal, this is the design. 

@bandinib-amzn would you update the title.

**Comment by ashwin-pc** [comment]
@seraphjiang thanks for clarifying that but it still does not address what's the need to merge this before the implementation is complete? To reiterate, I'm not against adding design documentation, but want to know the benefit of adding it before the actual implementation itself

**Comment by seraphjiang** [comment]
@ashwin-pc 
regardless open source, design should always go before implementation. That is common sense and engineering practices. If we don't have confidence on design, we could keep discussion.

About benefit, are we talking about we need to justify the benefit why design should proposed, reviewed, signed off before implementation? Design could be risk, so it should wait for implementation to prove?

**Comment by seanneumann** [comment]
TL;DR - 1) ideally you create design proposals in an Issue; 2) it won't hurt anything by merging this PR (I'm okay to merge).

Generally speaking, If you have a design proposal, I recommend to start by creating a GitHub Issue to discuss and gather feedback on your proposal. This is a great way to initiate discussions and can engage with the community, gather their thoughts and suggestions, and refine your design document based on the feedback received.

Once you have received sufficient feedback and iterated on your design document, you can then consider submitting a PR along with implementation changes. By going through the process of creating an issue first, you can ensure that your proposal aligns with the project's goals, discuss its feasibility, and address any concerns or alternative approaches that may arise. This collaborative approach promotes transparency and encourages open discussion before investing time and effort in implementing changes.

BUT... you went about it another way and I don't think it will cause any harm. So I'm ok with the merge.

**Comment by seraphjiang** [comment]
Thanks @seanneumann for clarification. @ashwin-pc do you have more questions or concern to unblock the PR?

File: docs/saved_objects/img/proposed_saved_object_service_workflow.png
Comment (zengyan-amazon) [inline docs/saved_objects/img/proposed_saved_object_service_workflow.png:1]: the UML diagrams are a bit miss leadings, the repository factory provider should only be called once during OSD bootstrap, at that time OSD create the repository instance, and use the repository instance to create saved object client. When user request coming in, the repository factory provider should not be involved.

**Comment by ashwin-pc** [comment]
@seanneumann it does hurt though. I have never seen a design proposal merged into a codebase before the design and i believe its for good reason. When the feature hasnt been implemented yet, it gives the user a false impression of whats in the codebase at that point in time. I often use these designs in code to understand what the state of the codebase is for the point in history that im looking at. If I see a saved object design that uses multiple datasources, I assume that the feature being described exists. And this is a workflow i have often used in this and other open source codebases. 

Also you mention that you are okay merging this, which i am too. My question here is specifically around its timing. Do see any particular reason for merging this early given the risks i have outlined?

**Comment by ashwin-pc** [comment]
@seraphjiang can you provide me any examples of the design being merged in before the implementation?

**Comment by seraphjiang** [comment]
> For a mature piece of software, merging in an uncooked feature whose "[d]esign ... need[s] iteration" will break our promise of semver; I agree with Ashwin that the code is not the place to do this when the software is not in v0.
> 
> I also don't agree with Sean's statement that "it won't hurt anything." Incomplete features will confuse users and burden the developers and contributors with potential bugs. It will also hurt the image of the software for having incomplete features and being buggy. Users who attempt to use the incomplete feature will be disappointed, resulting in a negative sentiment towards the feature. Even if the feature evolves into a perfect solution, the previous infamy will always be attached to it.
> 
> As part of refocusing, we decided that nothing incomplete will go into `main`, that every feature will reside in its own branch, and only after the feature development is complete and appropriate unit and integration tests are built (and pass), and with sufficient sign-offs, will it get to raise a PR.

@AMoo-Miki would you add input what part in this design make you think it is `incomplete feature` .

**Comment by AMoo-Miki** [comment]
> @AMoo-Miki would you add input what part in this design make you think it is `incomplete feature` .

Just the comment that said the design needs iteration; i haven't gotten a chance to review it yet.

**Comment by zengyan-amazon** [comment]
I think the purpose for this PR is to having the maintainers to review this design proposal, and we can iterate on it till we have an alignment. Unfortunately we currently do not have a well defined process for such kind of design review and design doc maintenance process, which we should workout one. I think we have following high level options:
1. Do the design review as PR (like this one), get it approved and merged in, then we follow up with implementation in separate PRs
1. Do the design review as PR, get it approved by maintainers (in the format as comments, not GitHub PR approvals), then publish the implementation in the same PR and merge design doc and implementation in at the same time.
   - I personally do not recommend this approach, as requires one single PR, and sometimes it will result in big PR that is hard to review and merge(resolve conflict overhead etc.)
1. Not check in design docs into the repo, and always use GitHub issues for design review. We will need to have clearer process for the review and approval though.

File: docs/saved_objects/img/proposed_saved_object_service_workflow.png
Comment (bandinib-amzn) [reply docs/saved_objects/img/proposed_saved_object_service_workflow.png:1]: Modified UML diagram.

**Comment by ashwin-pc** [comment]
@zengyan-amazon There is also the approach we follow today which i think we should not dismiss without good reason since its whats followed across the org. Design as an issue, then implementation as a PR with the agreed upon design in the same PR. And maybe this is something we can codify in the `.github` repo so that we aren't following different processes across the org.

@seraphjiang I also sense an urgency in the PR, are we blocking anything by not merging this in ASAP? Based on my conversation with @bandinib-amzn, it looks like she is not blocked on the implementation because of this, since this is just the design that folks have already agreed to. Also, based on the feedback it looks like we need a discussion here before we can all agree on the preferred approach, so I want to make sure that this isnt blocking any planned deliverables in the meantime.

**Comment by seraphjiang** [comment]
@ashwin-pc can you provide me any examples of the design being merged in before the implementation which hurt the user?

**Comment by seraphjiang** [comment]
> > @AMoo-Miki would you add input what part in this design make you think it is `incomplete feature` .
> 
> Just the comment that said the design needs iteration; i haven't gotten a chance to review it yet.

Thanks @AMoo-Miki , would you spend time to review the design. I do expect your input and approval.

**Review by zhongnansu — APPROVED** [review]

**Comment by seraphjiang** [comment]
@ashwin-pc @AMoo-Miki @zengyan-amazon @zhongnansu 
Could we focus on design review and get it approval than spending energy on process discussion? Could we expect team to spend quality of time to review and be responsible for the approval before implementation.

**Review by zengyan-amazon — APPROVED** [review]
